library/std/src/sys/uefi/common.rs

```diff
@@ -0,0 +1,477 @@
+    fn inner(
+        guid: &mut Guid,
+        boot_services: NonNull<r_efi::efi::BootServices>,
+        buf_size: &mut usize,
+        buf: *mut r_efi::efi::Handle,
+    ) -> io::Result<()> {
+        let r = unsafe {
+            ((*boot_services.as_ptr()).locate_handle)(
+                r_efi::efi::BY_PROTOCOL,
+                guid,
+                crate::ptr::null_mut(),
+                buf_size,
+                buf,
+            )
+        };
+
+        if r.is_error() { Err(status_to_io_error(r)) } else { Ok(()) }
+    }
+
+    let boot_services = uefi::env::get_boot_services().ok_or(BOOT_SERVICES_ERROR)?;
+    let mut buf_len = 0usize;
+
+    match inner(&mut guid, boot_services, &mut buf_len, crate::ptr::null_mut()) {
+        Ok(()) => unreachable!(),
+        Err(e) => match e.kind() {
+            io::ErrorKind::FileTooLarge => {}
+            _ => return Err(e),
+        },
+    }
+
+    // The returned buf_len is in bytes
+    let mut buf: Vec<r_efi::efi::Handle> =
+        Vec::with_capacity(buf_len / crate::mem::size_of::<r_efi::efi::Handle>());
+    match inner(&mut guid, boot_services, &mut buf_len, buf.as_mut_ptr()) {
+        Ok(()) => {
+            // SAFETY: This is safe because the call will succeed only if buf_len >= required
+            // length. Also, on success, the `buf_len` is updated with the size of bufferv (in
+            // bytes) written
+            unsafe { buf.set_len(buf_len / crate::mem::size_of::<r_efi::efi::Handle>()) };
+            Ok(buf.iter().filter_map(|x| NonNull::new(*x)).collect())
+        }
+        Err(e) => Err(e),
+    }
+}
+
+/// Open Protocol on a handle
+/// Implemented using `EFI_BOOT_SERVICES.OpenProtocol()`
+pub(crate) fn open_protocol<T>(
+    handle: NonNull<crate::ffi::c_void>,
+    mut protocol_guid: Guid,
+) -> io::Result<NonNull<T>> {
+    let boot_services = uefi::env::get_boot_services().ok_or(BOOT_SERVICES_ERROR)?;
+    let system_handle = uefi::env::get_system_handle().ok_or(SYSTEM_HANDLE_ERROR)?;
+    let mut protocol: MaybeUninit<*mut T> = MaybeUninit::uninit();
+
+    let r = unsafe {
+        ((*boot_services.as_ptr()).open_protocol)(
+            handle.as_ptr(),
+            &mut protocol_guid,
+            protocol.as_mut_ptr().cast(),
+            system_handle.as_ptr(),
+            crate::ptr::null_mut(),
+            r_efi::system::OPEN_PROTOCOL_GET_PROTOCOL,
+        )
+    };
+
+    if r.is_error() {
+        Err(status_to_io_error(r))
+    } else {
+        NonNull::new(unsafe { protocol.assume_init() })
+            .ok_or(io::error::const_io_error!(io::ErrorKind::Other, "Null Protocol"))
+    }
+}
+
+pub(crate) fn status_to_io_error(s: r_efi::efi::Status) -> io::Error {
+    use io::ErrorKind;
+    use r_efi::efi::Status;
+
+    // Keep the List in Alphabetical Order
+    match s {
+        Status::ABORTED => {
+            const_io_error!(ErrorKind::ConnectionAborted, "EFI_ABORTED")
```

Comment:
Well, I am using the UEFI Status name in the message because a Status code will have a different meaning (or several possible meanings) depending on the function that is being called. I'm not quite sure what a more human-readable message can look like.

Reply:
The UEFI specification contains descriptions for the error codes (see Appendix D), I think those would be fitting.